### DISCUSSION TITLE
Can't manage to setup basic auth properly HAProxy on pfsense

### DISCUSSION THREAD

**Discussion by schklom** [authelia/authelia#3898] [Q&A] [answered]
### Description

Basic auth doesn't work for me

### Expected Behaviour

Basic auth works

### Reproduction Steps

This is the part relevant to basic-auth that I have. Please tell me if you spot issues. The page I want to work is `https://apprise.mydomain:12345` :P
configuration.yaml
```yaml
access_control:
  default_policy: deny
  rules:
    - domain:
      - auth.mydomain
      - policy: bypass

    - domain:
      - apprise.mydomain
      policy: one_factor
      subject:
      - 'group:admin'
```

haproxy.cfg (PfSense setup via GUI), differences in relevant basic-auth parts from official configuration
```diff
global
	maxconn			10000
	log			/var/run/log	local0	debug
	stats socket /tmp/haproxy.socket level admin  expose-fd listeners
	uid			80
	gid			80
	nbthread			1
	hard-stop-after		15m
	chroot				/tmp/haproxy_chroot
	daemon
        ssl_stuff...
	server-state-file /tmp/haproxy_server_state
	lua-load		/var/etc/haproxy/luascript_json.lua
	lua-load		/var/etc/haproxy/luascript_auth-request.lua
	# https://www.authelia.com/docs/deployment/supported-proxies/haproxy.html
	lua-prepend-path /usr/local/share/lua/5.3/haproxy-lua-http.lua
	# https://www.haproxy.com/blog/accelerate-your-apis-by-using-the-haproxy-cache/
	cache mycache
	    total-max-size 1000     # MB (1 GB)
	    max-object-size 2000000 # bytes (2 MB)
	    max-age 86400           # seconds (1 day)

frontend fWAN
option forwardfor

- acl protected-frontends-basic hdr(host) -m reg -i ^(?i)(heimdall)\.example\.com
+ acl protected-frontends-basic var(txn.txnhost) -m str apprise.mydomain:12345

http-request set-var(txn.host) hdr(Host)

http-request set-var(req.scheme) str(https) if { ssl_fc }
http-request set-var(req.scheme) str(http) if !{ ssl_fc }

http-request set-var(req.questionmark) str(?) if { query -m found }

# Required headers
http-request set-header X-Real-IP %[src]
- http-request set-header X-Forwarded-Proto %[var(req.scheme)]
+ acl https ssl_fc
+ http-request set-header		X-Forwarded-Proto http if !https
+ http-request set-header		X-Forwarded-Proto https if https
http-request set-header X-Forwarded-Host %[req.hdr(Host)]
http-request set-header X-Forwarded-Uri %[path]%[var(req.questionmark)]%[query]

# Force `Authorization` header via query arg to /api/verify
http-request lua.auth-request be_authelia /api/verify?auth=basic if protected-frontends-basic

# Send 401 and pass `WWW-Authenticate` header on protected-frontend-basic if not pre-authenticated
http-request set-var(txn.auth) var(req.auth_response_header.www_authenticate) if protected-frontends-basic !{ var(txn.auth_response_successful) -m bool }
- http-response deny deny_status 401 hdr WWW-Authenticate %[var(txn.auth)] if { var(txn.host) -m reg -i ^(?i)(heimdall)\.example\.com } !{ var(txn.auth_response_successful) -m bool }
+ http-response deny deny_status 401 hdr WWW-Authenticate %[var(txn.auth)] if { var(txn.host) -m reg -i ^(?i)(apprise)\.mydomain:12345 } !{ var(txn.auth_response_successful) -m bool }

acl Apprise var(txn.txnhost) -m str apprise).mydomain:12345
use_backend BE_Apprise_with_Authelia_ipvANY  if  Apprise





backend BE_Apprise_with_Authelia_ipvANY
	mode			http
	id			152
	log			global
	timeout connect		10000
	timeout server		30000
	retries			3
	load-server-state-from-file	global
	compression algo gzip
	compression type text/html text/plain text/css
	# https://www.haproxy.com/blog/preserve-source-ip-address-despite-reverse-proxies/
	
	# https://www.haproxy.com/blog/protect-your-web-server-against-slowloris/
	timeout queue 60s
	option httpclose
	option abortonclose
	timeout http-request 5s
	
	# https://www.haproxy.com/blog/accelerate-your-apis-by-using-the-haproxy-cache/
	# Get from cache / put in cache
	acl	WeCanSafelyCacheThatFile	var(txn.txnpath) -m end -i .js .svg .png .css .xml
	http-request cache-use mycache if WeCanSafelyCacheThatFile
	http-response cache-store mycache if WeCanSafelyCacheThatFile
	
	acl			remote_user_exist	var(req.auth_response_header.remote_user) -m found
	acl			remote_groups_exist	var(req.auth_response_header.remote_groups) -m found
	acl			remote_name_exist	var(req.auth_response_header.remote_name) -m found
	acl			remote_email_exist	var(req.auth_response_header.remote_email) -m found
	http-request set-var(txn.txnpath) path
	http-request set-header Remote-User %[var(req.auth_response_header.remote_user)]  if  remote_user_exist 
	http-request set-header Remote-Groups %[var(req.auth_response_header.remote_groups)]  if  remote_groups_exist 
	http-request set-header Remote-Name %[var(req.auth_response_header.remote_name)]  if  remote_name_exist 
	http-request set-header Remote-Email %[var(req.auth_response_header.remote_email)]  if  remote_email_exist 
	server			Raspberry_Pi 172.16.1.11:20000 id 101  
```

### Additional Information

2-factor with TOTP works well.

Any ideas?

Logs when going to `https://apprise.mydomain:12345`
```json
{
    "level": "error",
    "method": "HEAD",
    "msg": "Error caught when verifying user authorization: basic auth requested via query arg, but no value provided via Authorization header",
    "path": "/api/verify",
    "remote_ip": "172.16.1.10",
    "stack": [{
            "File": "github.com/authelia/authelia/v4/internal/handlers/handler_verify.go",
            "Line": 463,
            "Name": "VerifyGET.func1"
        }, {
            "File": "github.com/authelia/authelia/v4/internal/middlewares/bridge.go",
            "Line": 54,
            "Name": "(*BridgeBuilder).Build.func1.1"
        }, {
            "File": "github.com/authelia/authelia/v4/internal/middlewares/headers.go",
            "Line": 16,
            "Name": "SecurityHeaders.func1"
        }, {
            "File": "github.com/fasthttp/router@v1.4.11/router.go",
            "Line": 414,
            "Name": "(*Router).Handler"
        }, {
            "File": "github.com/valyala/fasthttp@v1.38.0/http.go",
            "Line": 153,
            "Name": "(*Response).StatusCode"
        }, {
            "File": "github.com/valyala/fasthttp@v1.38.0/server.go",
            "Line": 2308,
            "Name": "(*Server).serveConn"
        }, {
            "File": "github.com/valyala/fasthttp@v1.38.0/workerpool.go",
            "Line": 224,
            "Name": "(*workerPool).workerFunc"
        }, {
            "File": "github.com/valyala/fasthttp@v1.38.0/workerpool.go",
            "Line": 196,
            "Name": "(*workerPool).getCh.func1"
        }, {
            "File": "runtime/asm_arm64.s",
            "Line": 1263,
            "Name": "goexit"
        }
    ],
    "time": "2022-08-25T23:39:37+02:00"
}{
    "level": "info",
    "method": "HEAD",
    "msg": "Access to https://apprise.mydomain:12345/ is not authorized to user \u003canonymous\u003e, sending 401 response with basic auth header",
    "path": "/api/verify",
    "remote_ip": "172.16.1.10",
    "time": "2022-08-25T23:39:37+02:00"
}

```

**Comment by james-d-elliott** [comment]
If you don't supply the `Authorization` header it will error like that. Provide the header with the basic scheme and it will work. If this is not sufficient to figure out the issue then please provide more than "doesn't work" to describe the issue.

**Comment by schklom** [comment]
> Provide the header with the basic scheme

Isn't that already done?
```cfg
http-request set-var(txn.host) hdr(Host)
```
and
```cfg
# Force `Authorization` header via query arg to /api/verify
http-request lua.auth-request be_authelia /api/verify?auth=basic if protected-frontends-basic

# Send 401 and pass `WWW-Authenticate` header on protected-frontend-basic if not pre-authenticated
http-request set-var(txn.auth) var(req.auth_response_header.www_authenticate) if protected-frontends-basic !{ var(txn.auth_response_successful) -m bool }
http-response deny deny_status 401 hdr WWW-Authenticate %[var(txn.auth)] if { var(txn.host) -m reg -i ^(?i)(apprise)\.mydomain:12345 } !{ var(txn.auth_response_successful) -m bool }
```
Or do you mean that I forgot something else?

**Reply by james-d-elliott** [reply]
That configures HAProxy to appropriately respond when Authelia sends the 401, by replying with the `WWW-Authenticate` header and a 401 response instead of redirecting the user. The client requesting the resource *MUST* provide a Authorization header and encode credentials with the basic scheme.

**Reply by schklom** [reply]
That was a quick reply, thanks! :)

> The client requesting the resource MUST provide a Authorization header and encode credentials with the basic scheme

I do not receive any prompt to authenticate on the browser, just a blank page when I visit `https://apprise.mydomain:12345`.
My experience with basic auth is from Traefik and other standalone services where the user can type username+password in a special window within the web browser, like below
![image](https://user-images.githubusercontent.com/23263086/186898593-da414095-6429-4feb-9bc2-aa9053370d24.png)

Is there a misunderstanding on my part? Does Authelia's basic auth work differently?

`curl -u username:password https://apprise.mydomain:12345` seems to work though.

Last, the doc at https://www.authelia.com/configuration/security/access-control/#one_factor says
> This policy requires the user at minimum complete 1FA successfully (username and password). This means if they have performed 2FA then they will be allowed to access the resource.

But the following happens:
1. Authenticate successfully with 2FA
2. Go to `https://apprise.mydomain:12345` with the web browser
3. Same result as if I had not authenticated with 2FA (unable to access the page), and the same Authelia logs as above come up.

**Reply by schklom** [reply]
The end result I want is this:
- `https://apprise.mydomain:12345/api` protected with basic-auth (even if it cannot be accessed with the web-browser)
- `https://apprise.mydomain:12345` protected with 2fa

Ideally I want to manage this entirely within Authelia. Is this currently possible, or do I need to use HAProxy to tag `https://apprise.mydomain:12345/api` with `protected-frontend-basic` and `https://apprise.mydomain:12345/` with `protected-frontend`?

**Reply by james-d-elliott** [reply]
1. Your understanding is correct as to what should be shown. The login box is shown if the WWW-Authenticate header is sent with the 401 response, which Authelia is definitely doing in your case. As to what part of this HAProxy is not forwarding, I am not sure, check the response when this occurs maybe.
2. When using the forced basic auth session auth is not possible at this time.
3. You will have to handle this in HAProxy and treat the two endpoints differently.

**Comment by schklom** [comment, marked as answer]
It turns out that I had set
```cfg
http-request redirect location https://auth.example.com/?rd=%[var(req.scheme)]://%[base]%[var(req.questionmark)]%[query] if protected-frontends !{ var(txn.auth_response_successful) -m bool }
# Send 401 and pass `WWW-Authenticate` header on protected-frontend-basic if not pre-authenticated
http-request set-var(txn.auth) var(req.auth_response_header.www_authenticate) if protected-frontends-basic !{ var(txn.auth_response_successful) -m bool }
http-response deny deny_status 401 hdr WWW-Authenticate %[var(txn.auth)] if { var(txn.host) -m reg -i ^(?i)(apprise)\.mydomain:12345 } !{ var(txn.auth_response_successful) -m bool }
```
in the wrong order.
Once I set them in the correct order, it worked!

Thank you so much for helping me figuring it out @james-d-elliott ! :+1: 

Solution:
* *hardcore*: basic auth for a subfolder (`/api`), it is a bit difficult as I had to introduce an ACL conditioned on an DIY ACL (`acl API ... /api` and `http-request set-var(txn.dontuseauthelia) str(true) if API` and use it as an ACL with `if { var(txn.dontuseauthelia) -m found }` or `if !{ var(txn.dontuseauthelia) -m found }`).
* *simpler*: Then I gave up because it would be a pain to maintain later, and am now using the much simpler `apprise-api.mydomain:12345/notify` with basic auth and `apprise.mydomain:12345` with 2FA.

**Reply by james-d-elliott** [reply]
Nice! It's a bit unclear what you actually did. Was this wrong in the docs as well? If so I'd like to correct it.

**Reply by schklom** [reply]
> It's a bit unclear what you actually did

I need basic auth only when the path is `/notify`, and 2FA otherwise. HAProxy requires a trick to do this and I don't like it. Since Authelia cannot mix the two by itself, I decided to have a subdomain (`apprise-api.mydomain:12345`) dedicated to basic auth with the `/notify` path, and the normal subdomain `apprise.mydomain:12345` is protected with 2FA. This is imo much cleaner.

> Was this wrong in the docs as well?

The docs are right, but I had re-ordered the config lines in order to be nicely organized with my other config instructions and did not realize the order was important. It was my mistake :P

If you want to avoid this kind of mistake for newbies like me, you could mention at the top of the config page that the order of the configuration instructions matters.

**Reply by james-d-elliott** [reply]
> I need basic auth only when the path is `/notify`, and 2FA otherwise. HAProxy requires a trick to do this and I don't like it. Since Authelia cannot mix the two by itself, I decided to have a subdomain (`apprise-api.mydomain:12345`) dedicated to basic auth with the `/notify` path, and the normal subdomain `apprise.mydomain:12345` is protected with 2FA. This is imo much cleaner.

The limitation we're facing here is both the integration method HAProxy supports and nginx supports do not allow us to do anything fancy. It'd be nice if they did. Unlike Traefik/Caddy/Skipper/Envoy which send basically the exact response Authelia responds with, HAProxy and nginx force us to craft the response in the proxy itself. HAProxy can be integrated with SPOE which I think would give us the ability to do more in this area, but that's an epic project.

> If you want to avoid this kind of mistake for newbies like me, you could mention at the top of the config page that the order of the configuration instructions matters.

Brilliant! Thanks for the feedback.

**Reply by schklom** [reply]
I am now facing a weird related problem (tell me if you prefer I make a new GitHub issue).

Apprise is a self-hosted service with an API to send notifications on e.g. Signal/Matrix/Gotify/etc.
The problem is that `curl` commands to the basic auth protected API trigger a notification despite having no/incorrect username+password.

Did I do something wrong, or is this a bug?

Here is my current setup (minimal example without much constraints):
* haproxy: `apprise-api.mydomain:12345` is behind basic auth with Authelia
* Authelia `configuration.yaml`:
```yaml
access_control:
  default_policy: deny
  rules: 
  - domain:
    - authelia.mydomain
    policy: bypass
  
  # Apprise-API (basic auth)
  - domain: apprise-api.mydomain
    policy: one_factor
    subject:
      - "group:admin"
```
and there is a user properly defined in the admin group.

Web browser `apprise-api.mydomain:12345` -> works as expected, with login window.
curl commands to `apprise-api.mydomain:12345` return `www-authenticate` but still trigger notifications on the backend.
```sh
$ curl -d 'tag=my_test_tag&body=test message' https://apse-api.mydomain:12345/notify/apprise -v
*   Trying <IP_ADDRESS>:12345...
* Connected to apse-api.mydomain (<IP_ADDRESS>) port 12345 (#0)
...
TLS stuff...
...
> POST /notify/apprise HTTP/2
> Host: apse-api.mydomain:12345
> user-agent: curl/7.81.0
> accept: */*
> content-length: 32
> content-type: application/x-www-form-urlencoded
...
TLS stuff...
...
< HTTP/2 401 
< www-authenticate: Basic realm="Authentication required"
< content-length: 0
< 
* Connection #0 to host apse-api.mydomain left intact
```

**Reply by nightah** [reply]
That response indicates you need to pass in the username and password for Basic Authentication and unless you've redacted your curl command it isn't present.

```
       -u, --user <user:password>
              Specify the user name and password to use for server authentication. Overrides --netrc and --netrc-optional.

              If you simply specify the user name, curl will prompt for a password.

              The user name and passwords are split up on the first colon, which makes it impossible to use a colon in the user name with this option. The password can, still.

              On  systems  where it works, curl will hide the given option argument from process listings. This is not enough to protect credentials from possibly getting seen by other users on the same system as they will still be visible for a moment before cleared. Such sensitive data should be retrieved
              from a file instead or similar and never used in clear text in a command line.

              When using Kerberos V5 with a Windows based server you should include the Windows domain name in the user name, in order for the server to successfully obtain a Kerberos Ticket. If you do not, then the initial authentication handshake may fail.

              When using NTLM, the user name can be specified simply as the user name, without the domain, if there is a single domain and forest in your setup for example.

              To specify the domain name use either Down-Level Logon Name or UPN (User Principal Name) formats. For example, EXAMPLE\user and user@example.com respectively.

              If you use a Windows SSPI-enabled curl binary and perform Kerberos V5, Negotiate, NTLM or Digest authentication then you can tell curl to select the user name and password from your environment by specifying a single colon with this option: "-u :".

              If this option is used several times, the last one will be used.

              Example:
               curl -u user:secret https://example.com
```

**Reply by schklom** [reply]
@nightah I know, this is the weird part.
Even with that response, the backend is triggered and sends me a notification (backend is an API to send notifications).
It seems the curl message goes through, despite the response and lack of credentials. Providing any username+password changes the response a little, but in both cases the backend gets triggered (bug or misconfiguration).
